telethon/network/mtprotosender.py

```diff
@@ -386,20 +386,27 @@ def _start_reconnect(self):
             data = self._state.encrypt_message_data(data)
             try:
                 await self._connection.send(data)
-            except ConnectionError:
-                self._log.info('Connection closed while sending data')
-                self._start_reconnect()
-                return
+            except ConnectionError as e:
+                __log__.error('Connection closed while sending data')
+                # TODO we will reconnect in _recv loop
+                # this will cause double reconnection
+                # self._start_reconnect()
+                error = e
 
             for state in batch:
                 if not isinstance(state, list):
                     if isinstance(state.request, TLRequest):
-                        self._pending_state[state.msg_id] = state
+                        if not error:
+                            self._pending_state[state.msg_id] = state
+                        else:
+                            state.future.set_exception(error)
                 else:
                     for s in state:
                         if isinstance(s.request, TLRequest):
-                            self._pending_state[s.msg_id] = s
-
+                            if not error:
+                                self._pending_state[s.msg_id] = s
+                            else:
+                                s.future.set_exception(error)
```

Comment:
Maybe consider something like this?
```python3
                 if not isinstance(state, list):
                     state = [state]
                 for s in state:
                     if isinstance(s.request, TLRequest):
                         if not error:
                             self._pending_state[s.msg_id] = s
                         else:
                             s.future.set_exception(error)
```

Reply:
I think we should refactor `send_queue` a bit instead. Don't make it return a list of either states or lists, but rather just a list of states. We don't make use of that nesting here anyway. But perhaps that should be for another PR.

Also note that your suggestion should be applied to `s` not `state`.